### PR TITLE
Encodings: Ascii, Utf8, Hex and Base64

### PR DESCRIPTION
**Pull request by jdonaghue**

### CODE REVIEW

File: src/encoding.ts [hunk 1]
```diff
@@ -0,0 +1,379 @@
...
+		toJSON: function (): string {
+			return JSON.stringify(Array.prototype.slice.call(this));
+		},
+		length: data.length
+	}];
+}
+
+function validateDecodingArgs(data: any) {
+	if (data == null) {
+		return false;
```
Comment (kitsonk) [inline src/encoding.ts]: Why not `===`?

File: src/encoding.ts [hunk 2]
```diff
@@ -0,0 +1,379 @@
...
+	}
+
+	if (data.length == null) {
+		throw new TypeError('Argument must have a length property');
```
Comment (kitsonk) [inline src/encoding.ts]: Won't this throw weird... should it be `if ('length' in data) {`?

File: src/encoding.ts (hunk 1 above)
Comment (jdonaghue) [inline src/encoding.ts]: I opted to use == here so that it will catch undefined as well

File: src/encoding.ts [hunk 3]
```diff
@@ -0,0 +1,379 @@
...
+		let decoded = '';
+
+		for (let i = 0, length = data.length; i < length; i++) {
+			let byte1 = data[i] & 0xFF;
+
+			if ((byte1 & 0x80) === 0) {
+				decoded += decodeUtf8EncodedCodePoint(byte1);
+			}
+			else if ((byte1 & 0xE0) === 0xC0) {
+				let byte2 = data[++i] & 0xFF;
```
Comment (kitsonk) [inline src/encoding.ts:168]: Why are these being hoisted?  Wouldn't it be better to declare them above instead of conditional hoisting?

File: src/encoding.ts [hunk 4]
```diff
@@ -0,0 +1,379 @@
...
+	 *
+	 * @param data
+	 * The byte buffer to be decoded
+	 */
+	static decode(data: ByteBuffer): string {
+		if (!validateDecodingArgs(data)) {
+			return '';
+		}
+
+		let decoded = '';
```
Comment (kitsonk) [inline src/encoding.ts:159]: I looked at the style guide, and it doesn't address this sort of implicit typing, but I don't like it.  I would much rather we were explicitly typing wherever reasonable, because it makes intent so much easier.

File: tests/unit/encoding.ts [hunk 5]
```diff
@@ -0,0 +1,251 @@
...
+	{
+		encoding: '\uDFFF',
+		decoding: [0xED, 0xBF, 0xBF]
+	}
+];
+
+registerSuite({
+	name: 'Encoding Classes',
+
+	'Ascii': {
```
Comment (kitsonk) [inline tests/unit/encoding.ts]: No explicit tests for `validateDecodingArgs()`.  We should try to have negative tests as well as positive tests, especially when there is code that is specifically validating arguments, we should test for invalid arguments.

File: src/encoding.ts (hunk 2 above)
Comment (kfranqueiro) [inline src/encoding.ts]: If there's no expectation that data would be explicitly null or undefined (only not defined at all), then that would make more sense here.

File: src/encoding.ts (hunk 3 above)
Comment (kfranqueiro) [inline src/encoding.ts:168]: What do you mean by "hoisted"?  `let` has block scope.

File: src/encoding.ts (hunk 4 above)
Comment (kfranqueiro) [inline src/encoding.ts:159]: Good point about it not being in the style guide, but it was a conscious agreement during earlier discussions.

I don't really see the point to `let decoded: string = '';`, since you can obviously tell that `''` is already a string.

The one possible instance where I'd consider enforcing explicit typing even where unnecessary is for return types of exported functions.

File: src/encoding.ts (hunk 3 above)
Comment (kitsonk) [inline src/encoding.ts:168]: I didn't realise ES6 let was block scope.  Learned something new.  Though obviously the emit to ES5, this wouldn't be block scope though, right?  Only when emitted ES6.  I guess TS would do any guarding against mis-scoping?

File: src/encoding.ts (hunk 4 above)
Comment (kitsonk) [inline src/encoding.ts:159]: Fair enough it was agreed... maybe adding to the style guide that explicit agreement.

I would still argue that `let decode: string = '''` still makes the intent explicitly clear, and you can only get away with the with primitives (IIRC).

File: src/encoding.ts (hunk 3 above)
Comment (kfranqueiro) [inline src/encoding.ts:168]: Yeah, that's the idea.  TS also ensures that variable names are unique if they need to be (e.g. if you have one block-scoped `i` variable shadowing another one in an outer block scope).

File: src/encoding.ts (hunk 4 above)
Comment (kfranqueiro) [inline src/encoding.ts:159]: I have a local change to the style guide to add this, though I'm hoping we merge the contributing and style guides into one repo rather than having to update it in 4 or more of them.

Actually type inferencing happens on objects too...and those are the times it _can_ bite you.  Primitives are where it seems particularly excessive.

File: src/encoding.ts (hunk 2 above)
Comment (jdonaghue) [inline src/encoding.ts]: Good suggestion, not sure why I wrote it the way I did

File: src/encoding.ts [hunk 6]
```diff
@@ -0,0 +1,379 @@
...
+	return true;
+}
+
+function validateUtf8EncodedCodePoint(codePoint: number): void {
+	if ((codePoint & 0xC0) !== 0x80) {
+		throw Error('Invalid continuation byte');
+	}
+}
+
+function decodeUtf8EncodedCodePoint(codePoint: number, validationRange: number[] = [0, Infinity], checkSurrogate?: boolean): string {
```
Comment (kfranqueiro) [inline src/encoding.ts]: Space after `[` and before `]` in default value

File: src/encoding.ts [hunk 7]
```diff
@@ -0,0 +1,379 @@
...
+	if (codePoint < validationRange[0] || codePoint > validationRange[1]) {
+		throw Error('Invalid continuation byte');
+	}
+
+	if (checkSurrogate && codePoint >= 0xD800 && codePoint <= 0xDFFF) {
```
Comment (kfranqueiro) [inline src/encoding.ts:17]: We were going to make these constants either here or in `string`, right?

File: src/encoding.ts [hunk 8]
```diff
@@ -0,0 +1,379 @@
...
+	toJSON?: () => string;
+	length: number;
+}
+
+/**
+ * Ascii codec that provides facilities for encoding a string (or string coerced object)
+ * into an Ascii encoded byte buffer as well as decoding an Ascii encoded byte buffer
+ * into a string.
+ */
+export class Ascii {
```
Comment (kfranqueiro) [inline src/encoding.ts]: - Is there a point to these being classes rather than objects with 2 functions?
- Should there be an interface that these implement/conform to?

File: src/encoding.ts [hunk 9]
```diff
@@ -0,0 +1,379 @@
...
+	 * @param data
+	 * Any object that is either a string or can be coerced into a string
+	 */
+	static encode(data: any): ByteBuffer {
+		let buffer: ByteBuffer;
+
+		[ data, buffer ] = normalizeEncodingArgs(data, Ascii);
+
+		for (let i = 0, length = data.length; i < length; i++) {
+			buffer[i] = data.charCodeAt(i);
```
Comment (kfranqueiro) [inline src/encoding.ts:59]: Is this technically accurate, given that `charCodeAt` returns a 16-bit numeric Unicode value?

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/charCodeAt

File: src/encoding.ts [hunk 10]
```diff
@@ -0,0 +1,379 @@
...
+	 * @param data
+	 * Any object that is either a string or can be coerced into a string
+	 */
+	static encode(data: any): ByteBuffer {
+		let buffer: ByteBuffer;
+
+		[ data, buffer ] = normalizeEncodingArgs(data, Hex);
+
+		for (let i = 0, length = data.length; i < length; i++) {
+			let encodedChar = parseInt(data.charCodeAt(i).toString(16), 16);
```
Comment (kfranqueiro) [inline src/encoding.ts]: Can you explain to an encoding noob (me) what the `parseInt` and `toString(16)` here accomplishes?  It appears to me that they'd just cancel each other out.

File: src/encoding.ts [hunk 11]
```diff
@@ -0,0 +1,379 @@
...
+
+		return decoded;
+	}
+}
+
+/**
+ * Base64 codec that provides facilities for encoding a string (or string coerced object)
+ * into an Base64 encoded byte buffer as well as decoding an Base64 encoded byte buffer
+ * into a string.
+ * Adapted from http://www.nczonline.net/blog/2009/12/08/computer-science-in-javascript-base64-encoding/
```
Comment (kfranqueiro) [inline src/encoding.ts]: This should actually be able to use `atob` and `btoa` in all but IE9 (possibly with [some extra considerations for Unicode](https://developer.mozilla.org/en-US/docs/Web/API/WindowBase64/btoa#Unicode_Strings)).

Meanwhile, Dojo 1 already has base64 code that we can lift from (though the code is unfortunately harder to read...)

**Comment by kfranqueiro** [comment]
Have we considered having tests that do encode -> decode and/or decode -> encode to ensure that the value is the same as the original after running through both operations?  (I'm not sure offhand if any of the existing tests are effectively doing that.)

File: src/encoding.ts [hunk 12]
```diff
@@ -0,0 +1,379 @@
...
+		encoded += String.fromCharCode(codePoint >>> 0x10 & 0x03FF | 0xD800);
+		codePoint = 0xDC00 | codePoint & 0x03FF;
+	}
+
+	encoded += String.fromCharCode(codePoint);
+
+	return encoded;
+}
+
+export interface ByteBuffer {
```
Comment (kfranqueiro) [inline src/encoding.ts]: @jason0x43 may have more detailed thoughts on this, but it seems like it'd be far more useful (though probably more difficult) to be able to return something that is useful in the respective platform this is used.

At the _very_ least, it seems like it'd be more useful for this to actually be an array, not just an array-like.

File: src/encoding.ts (hunk 9 above)
Comment (kfranqueiro) [inline src/encoding.ts:59]: dojox/encoding in Dojo 1 already has an `ascii85` module which looks quite different...  However, I'm not sure whether that's actually what we're looking for here.

https://en.wikipedia.org/wiki/Ascii85

File: src/encoding.ts [hunk 13]
```diff
@@ -0,0 +1,379 @@
...
+			}
+		}
+
+		return decoded;
+	}
+}
+
+/**
+ * Hex codec that provides facilities for encoding a string (or string coerced object)
+ * into an Hex encoded byte buffer as well as decoding an Hex encoded byte buffer
```
Comment (jason0x43) [inline src/encoding.ts:213]: The original goal was to convert a hex encoded string, like "DEADBEEF01", into a byte buffer, and vice versa, rather than to store a string in a particular binary format.

File: src/encoding.ts (hunk 8 above)
Comment (jdonaghue) [inline src/encoding.ts]: `Is there a point to these being classes rather than objects with 2 functions?`
There isn't really a great reason for these being classes over plain old objects. I opted for classes though because they can be extended. So, if someone were to want to subclass any of these they would be able to. I know thats not an amazing argument but when comparing the two options a class based approach is the more flexible of the two. That said I am not married to it and could easily be convinced that an object based approach is better

`Should there be an interface that these implement/conform to?`
There isn't an interface because all of the methods are static.

File: src/encoding.ts (hunk 7 above)
Comment (jdonaghue) [inline src/encoding.ts:17]: yes, thank you for reminding me!

File: src/encoding.ts (hunk 1 above)
Comment (kitsonk) [inline src/encoding.ts]: Well if you are going to test for any falsey value, why not `!data` then?

File: src/encoding.ts [hunk 14]
```diff
@@ -0,0 +1,347 @@
+const BASE64_KEYSTR = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/=';
+
+function normalizeEncodingArgs(data: any): [ string, number[] ] {
```
Comment (jason0x43) [inline src/encoding.ts]: Since we're not using a custom ByteBuffer class, this method doesn't seem to be accomplishing much. We should just have the encoding functions only accept strings.

File: src/encoding.ts [hunk 15]
```diff
@@ -0,0 +1,347 @@
...
+ * into a string.
+ */
+export class Ascii {
+	/**
+	 * Encode a string or string coerced object into an Ascii byte buffer
+	 *
+	 * @param data
+	 * Any object that is either a string or can be coerced into a string
+	 */
+	static encode(data: any): number[] {
```
Comment (jason0x43) [inline src/encoding.ts]: Is there any reason not to make the signature `encode(data: string)`?

File: src/encoding.ts [hunk 16]
```diff
@@ -0,0 +1,347 @@
...
+		return buffer;
+	}
+
+	/**
+	 * Decode an Ascii encoded byte buffer into a string
+	 *
+	 * @param data
+	 * The byte buffer to be decoded
+	 */
+	static decode(data: number[]): string {
```
Comment (jason0x43) [inline src/encoding.ts]: This would be more convenient as `decode(data: Uint8Array | Buffer | number[])` since Uint8Array and Buffer act the same as far as `decode` is concerned (so the same code path would work for all of them). You could use a type alias, like `type ByteBuffer = Uint8Array | Buffer | number[]`, giving `decode(data: ByteBuffer)`.

File: src/encoding.ts (hunk 4 above)
Comment (jason0x43) [inline src/encoding.ts:159]: As long as the compiler can unambiguously figure things out, is there any benefit to including unnecessary explicit types?

File: src/encoding.ts [hunk 17]
```diff
@@ -0,0 +1,347 @@
...
+	}
+}
+
+/**
+ * Utf8 codec that provides facilities for encoding a string (or string coerced object)
+ * into an Utf8 encoded byte buffer as well as decoding an Utf8 encoded byte buffer
+ * into a string.
+ * Inspired by the work of: https://github.com/mathiasbynens/utf8.js
+ */
+export class Utf8 {
```
Comment (jason0x43) [inline src/encoding.ts]: The codecs should implement a `Codec` interface.

File: src/encoding.ts (hunk 4 above)
Comment (kfranqueiro) [inline src/encoding.ts:159]: See number 2 under https://github.com/dojo/guidelines/blob/master/STYLE.md#typescript for how I worded it.  Basically I'm only advocating explicit types where necessary or in exported functions or class methods.

File: src/encoding.ts (hunk 17 above)
Comment (jdonaghue) [inline src/encoding.ts]: You can't have static methods on an interface in Typescript unfortunately.

File: src/encoding.ts (hunk 16 above)
Comment (jdonaghue) [inline src/encoding.ts]: ok good idea I will make this change

File: src/encoding.ts (hunk 8 above)
Comment (jason0x43) [inline src/encoding.ts]: I think the benefits of using 2-function modules with an interface outweigh the ability to subclass. Having an interface will allow these to be easily passed into functions, which will be a key use of codecs. If someone wants to extend one of these codecs, they can just wrap it and call the wrapped object's methods.

Comment (jdonaghue) [inline src/encoding.ts]: ok that sounds reasonable. I will make that change as well.

File: src/encoding.ts (hunk 4 above)
Comment (jason0x43) [inline src/encoding.ts:159]: Hmmm...I don't really think adding the type improves clarity much; you're going to know what it should be the first time you try to compile the code, and your IDE is always going to be able to tell you. It could probably help enforce the public API, though. Explicitly setting a return type will prevent you from accidentally returning something invalid (which will probably only come up in library code, but hey, that's what we're writing).

File: src/encoding.ts [hunk 18]
```diff
@@ -0,0 +1,337 @@
...
+	encode(data: string): number[];
+	decode(data: ByteBuffer): string;
+}
+
+/**
+ * Ascii codec that provides facilities for encoding a string
+ * into an Ascii encoded byte buffer as well as decoding an Ascii encoded byte buffer
+ * into a string.
+ */
+export const Ascii: Codec = {
```
Comment (jason0x43) [inline src/encoding.ts]: The codec names should use lower camel case now since they're not classes. (And they're not constants in the sense that they should be all-capped).

Comment (kfranqueiro) [inline src/encoding.ts]: Just to avoid any potential confusion here... you're not saying `const` is unwarranted here, right?  You're just saying that they shouldn't be all-caps either?

File: tests/unit/encoding.ts [hunk 19]
```diff
@@ -0,0 +1,235 @@
...
+		'.encode()'() {
+			let buffer = Ascii.encode(ENCODED_STRING);
+			assert.strictEqual(buffer.length, ENCODED_STRING.length);
+
+			for (let i = 0, length = buffer.length; i < length; i++) {
+				assert.strictEqual(buffer[i], ASCII_BUFFER[i]);
+			}
```
Comment (jason0x43) [inline tests/unit/encoding.ts]: Since we're outputting number arrays all the time now, the test could just use assert.deepEqual instead of checking the length and individual character values. Also, there's an assert.lengthOf for checking the length of things.

File: src/encoding.ts (hunk 4 above)
Comment (kfranqueiro) [inline src/encoding.ts:159]: Yeah, ensuring you don't accidentally return something unexpected is definitely part of why I put it in there.  Also just that it's generally easier to eyeball the first line of the function to get all the type information out of it if you're reading it.

File: tests/unit/encoding.ts [hunk 20]
```diff
@@ -0,0 +1,200 @@
...
+			assert.deepEqual(buffer, ASCII_BUFFER);
+
+			buffer = ascii.encode('1');
+			assert.lengthOf(buffer, 1);
+			assert.strictEqual(buffer[0], 49);
+		},
+
+		'.decode()'() {
+			let decoded = ascii.decode(ASCII_BUFFER);
+			assert.lengthOf(decoded, ENCODED_STRING.length);
```
Comment (jason0x43) [inline tests/unit/encoding.ts]: There's no need for a lengthOf check when the next thing is check for strict equality.

File: tests/unit/encoding.ts [hunk 21]
```diff
@@ -0,0 +1,200 @@
...
+	}
+];
+
+registerSuite({
+	name: 'Encoding Classes',
+
+	'ascii': {
+		'.encode()'() {
+			let buffer = ascii.encode(ENCODED_STRING);
+			assert.lengthOf(buffer, ENCODED_STRING.length);
```
Comment (jason0x43) [inline tests/unit/encoding.ts]: There's no need for a lengthOf check when the next thing is checking for deep equality (which includes length).

File: src/encoding.ts [hunk 22]
```diff
@@ -0,0 +1,341 @@
...
+	/**
+	 * Encode a string into an Ascii byte buffer
+	 *
+	 * @param data
+	 * A text string to be encoded
+	 */
+	encode(data: string): number[] {
+		let buffer: number[] = [];
+
+		for (let i = 0, length = data.length; i < length; i++) {
```
Comment (jason0x43) [inline src/encoding.ts:58]: There should be a check that data isn't null/undefined. If it is null, the encoder should just return an empty array.

File: src/encoding.ts [hunk 23]
```diff
@@ -0,0 +1,341 @@
...
+			 * Surrogates
+			 * http://en.wikipedia.org/wiki/Universal_Character_Set_characters
+			 */
+			if (encodedChar > HIGH_SURROGATE_MIN && encodedChar < HIGH_SURROGATE_MAX) {
+				let lowSurrogate = data.charCodeAt(++i);
+				if (lowSurrogate >= LOW_SURROGATE_MIN && lowSurrogate < LOW_SURROGATE_MAX) {
+					encodedChar = 0x010000 + (encodedChar - 0xD800) * 0x0400 + (lowSurrogate - 0xDC00);
+				}
+				else {
+					i--;
```
Comment (jason0x43) [inline src/encoding.ts]: Directly modifying the loop-control variable inside a for loop seems...odd. Maybe that's just me.

Comment (jdonaghue) [inline src/encoding.ts]: would it be preferable for me to change this to a while-loop?

Comment (kfranqueiro) [inline src/encoding.ts]: I think this would be acceptable as a `for` loop if you increment `i` only _inside_ the `if` (at which point you can remove the `else` entirely).  One control var tweak is better than two.  (You'd have to do `i + 1` on the line before the `if` though.)

A while loop wouldn't really help, since you'd have to explicitly increment twice in some cases.

File: src/encoding.ts [hunk 24]
```diff
@@ -0,0 +1,337 @@
+type ByteBuffer = Uint8Array | Buffer | number[];
+const BASE64_KEYSTR = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/=';
+const HIGH_SURROGATE_MIN = 0xD800;
```
Comment (kfranqueiro) [inline src/encoding.ts]: My original suggestion was to put these surrogate constants in `string.ts` then import them here (and also use them there).

File: src/encoding.ts (hunk 23 above)
Comment (jdonaghue) [inline src/encoding.ts]: I have made this change

File: src/encoding.ts [hunk 25]
```diff
@@ -0,0 +1,333 @@
...
+ * into a string.
+ */
+export const ascii: Codec = {
+	/**
+	 * Encode a string into an Ascii byte buffer
+	 *
+	 * @param data
+	 * A text string to be encoded
+	 */
+	encode(data: string = ''): number[] {
```
Comment (jason0x43) [inline src/encoding.ts]: This doesn't handle the case when data is null, only when it's undefined. Use something like `validateDecodingArgs` inside the function instead.

File: src/encoding.ts [hunk 26]
```diff
@@ -0,0 +1,333 @@
...
+		let buffer: number[] = [];
+
+		for (let i = 0, length = data.length; i < length; i++) {
+			let encodedChar = data.charCodeAt(i);
+
+			/**
+			 * Surrogates
+			 * http://en.wikipedia.org/wiki/Universal_Character_Set_characters
+			 */
+			if (encodedChar > HIGH_SURROGATE_MIN && encodedChar < HIGH_SURROGATE_MAX) {
```
Comment (jason0x43) [inline src/encoding.ts]: Should this range be inclusive (>= and <= instead of > and <)?

File: src/encoding.ts [hunk 27]
```diff
@@ -0,0 +1,333 @@
...
+				let lowSurrogate = data.charCodeAt(i + 1);
+				if (lowSurrogate >= LOW_SURROGATE_MIN && lowSurrogate < LOW_SURROGATE_MAX) {
```
Comment (jason0x43) [inline src/encoding.ts]: Should the `LOW_SURROGATE_MAX` comparison be inclusive?

File: src/encoding.ts [hunk 28]
```diff
@@ -0,0 +1,333 @@
...
+ */
+export const utf8: Codec = {
+	/**
+	 * Encode a string into an Utf8 byte buffer
+	 *
+	 * @param data
+	 * A text string to be encoded
+	 */
+	encode(data: string = ''): number[] {
+		let buffer: number[] = [];
```
Comment (kitsonk) [inline src/encoding.ts]: Ok, might be another thing with me getting my head wrapped around ES6, but why isn't this a `const`?  You are never going to reassign this or modify the pointer in any way.

File: src/encoding.ts (hunk 26 above)
Comment (jason0x43) [inline src/encoding.ts]: Actually, why not use `string.codePointAt` for this?

File: src/encoding.ts [hunk 29]
```diff
@@ -0,0 +1,333 @@
...
+	 */
+	decode(data: ByteBuffer): string {
+		if (!validateDecodingArgs(data)) {
+			return '';
+		}
+
+		let decoded = '';
+
+		for (let i = 0, length = data.length; i < length; i++) {
+			decoded += parseInt('' + data[i], 10).toString(16).toUpperCase();
```
Comment (jason0x43) [inline src/encoding.ts]: Why use `parseInt` instead of just `data[i].toString(16)`? `data[i]` should already be a number.

File: src/encoding.ts (hunk 26 above)
Comment (kfranqueiro) [inline src/encoding.ts]: RE using `codePointAt`, that was an idea I had brought up, but I wasn't sure if it'd add complexity regarding knowing how many characters to skip.

And yeah, I'd suspect this should be inclusive, not exclusive.

File: tests/unit/encoding.ts [hunk 30]
```diff
@@ -0,0 +1,204 @@
...
+	'ascii': {
+		'.encode()'() {
+			let buffer = ascii.encode(ENCODED_STRING);
+			assert.deepEqual(buffer, ASCII_BUFFER);
+
+			buffer = ascii.encode('1');
+			assert.lengthOf(buffer, 1);
+			assert.strictEqual(buffer[0], 49);
+
+			assert.deepEqual([], ascii.encode(undefined));
```
Comment (jason0x43) [inline tests/unit/encoding.ts:77]: Also test `null`

File: tests/unit/encoding.ts [hunk 31]
```diff
@@ -0,0 +1,204 @@
...
+		},
+
+		'.decode()'() {
+			let decoded = ascii.decode(ASCII_BUFFER);
```
Comment (jason0x43) [inline tests/unit/encoding.ts:82]: It'd be good to add tests using Buffers and Uint8Arrays (on compatible platforms) since a decoder can theoretically accept those as well as numeric arrays.

File: src/encoding.ts (hunk 28 above)
Comment (kfranqueiro) [inline src/encoding.ts]: I think that's a fair point here and was just overlooked.  Easy to miss those things with objects that get mutated but not reassigned.

File: src/encoding.ts (hunk 26 above)
Comment (jdonaghue) [inline src/encoding.ts]: The only problem with using `codePointAt` is that I need to keep track of what position I am at in the source string and that would be difficult with that utility function because it doesn't return any indicator of whether a low surrogate was provided or not. 

Also, according to this: http://people.mozilla.org/~jorendorff/es6-draft.html#sec-string.prototype.codepointat both of these ranges should both be exclusive which if thats correct then both this and `codePointAt` are wrong currently

File: src/encoding.ts (hunk 28 above)
Comment (jdonaghue) [inline src/encoding.ts]: yeah good point, i will change

File: src/encoding.ts (hunk 29 above)
Comment (jdonaghue) [inline src/encoding.ts]: I was using `parseInt` because it was originally returning a collection of `numbers`. Now it is returning a string so `parseInt` doesn't make sense. Good catch!

File: tests/unit/encoding.ts (hunk 31 above)
Comment (jdonaghue) [inline tests/unit/encoding.ts:82]: Good idea I will add some `Buffer` and `Uint8Array` tests, thanks!

File: src/encoding.ts (hunk 26 above)
Comment (jdonaghue) [inline src/encoding.ts]: duh! I read that wrong, you are right they should be inclusive. I will change

Comment (kfranqueiro) [inline src/encoding.ts]: IINM, I think you're reading the spec backwards.  Step 9 tests for things being _outside_ of the range, in which case the endpoints are exclusive.  We're testing for them being _inside_ the range here and in `string.ts`.  (The difference between `n > max || n < min`, and `n <= max && n >= min`)

File: tests/unit/encoding.ts [hunk 32]
```diff
@@ -0,0 +1,246 @@
...
+
+			assert.strictEqual(ascii.decode(undefined), '');
+			assert.strictEqual(ascii.decode(null), '');
+
+			if (has('host-node')) {
+				let buffer = new Buffer('The cat and the hat');
+				assert.strictEqual(ascii.decode(buffer), 'The cat and the hat');
+			}
+			else if (has('host-browser')) {
+				let buffer = new Uint16Array(ASCII_BUFFER);
```
Comment (jason0x43) [inline tests/unit/encoding.ts:95]: This fails in IE9. We should add a has test for typed arrays and use that here.

File: src/encoding.ts [hunk 33]
```diff
@@ -0,0 +1,349 @@
...
+export const ascii: Codec = {
+	/**
+	 * Encode a string into an Ascii byte buffer
+	 *
+	 * @param data
+	 * A text string to be encoded
+	 */
+	encode(data: string): number[] {
+		if (!validateArguments(data)) {
+			data = '';
```
Comment (jason0x43) [inline src/encoding.ts]: You could just return `[]` here.

File: tests/unit/encoding.ts [hunk 34]
```diff
@@ -0,0 +1,246 @@
...
+			assert.strictEqual(ENCODED_STRING, ascii.decode(ascii.encode(ENCODED_STRING)));
+
+			assert.strictEqual(ascii.decode(undefined), '');
+			assert.strictEqual(ascii.decode(null), '');
+
+			if (has('buffer')) {
+				let buffer = new Buffer('The cat and the hat');
+				assert.strictEqual(ascii.decode(buffer), 'The cat and the hat');
+			}
+			else if (has('uint16Array')) {
```
Comment (jason0x43) [inline tests/unit/encoding.ts:94]: Any particular reason to use uint16 here instead of uint8? The values in the test buffers all fit in 8-bits, and the Codec interface is defined as accepting a Uint8Array.

File: src/encoding.ts [hunk 35]
```diff
@@ -0,0 +1,349 @@
...
+ */
+export const ascii: Codec = {
+	/**
+	 * Encode a string into an Ascii byte buffer
+	 *
+	 * @param data
+	 * A text string to be encoded
+	 */
+	encode(data: string): number[] {
+		if (!validateArguments(data)) {
```
Comment (jason0x43) [inline src/encoding.ts]: Since `validateArguments` is just testing for data to be null or undefined, just say

``` js
if (data == null) {
   ..
}
```

and get rid of `validateArguments`.

File: tests/unit/encoding.ts (hunk 34 above)
Comment (jdonaghue) [inline tests/unit/encoding.ts:94]: Because the test string contains characters higher than an octect supports. 65399 and  65382 wont fit in 8-bits

File: src/encoding.ts (hunk 35 above)
Comment (jdonaghue) [inline src/encoding.ts]: good call, I will do this thanks!

File: tests/unit/encoding.ts (hunk 34 above)
Comment (kfranqueiro) [inline tests/unit/encoding.ts:94]: Isn't ASCII 8-bit?  Why would we be testing with 16-bit numbers?

Also, I'm not sure `has` tests are warranted for things that we're only using in tests, not implementation.

**Comment by kfranqueiro** [comment]
I have an [encoding-cleanup](https://github.com/kfranqueiro/core/commits/encoding-cleanup) branch which I hope to merge later today or tomorrow.  The main things addressed were:
- exporting the `ByteBuffer` type since other APIs will want to reference it (e.g. crypto)
- jsdoc comment tweaking
- removing the `has` tests this added (since they're only used in unit tests, not implementations)
- using separate strings for ascii and utf8 tests
- using Uint8Array where it makes sense rather than Uint16Array in tests
- replacing more values corresponding to the new surrogate constants

**Comment by kfranqueiro** [comment]
Merged in https://github.com/dojo/core/commit/8b70efadfaea3e67df52743dceb7b18af64fae80.
